roachprod: improve the way cockroach is run

RaduBerinde opened the pull request:
While running some stress tests with TPCH, I observed two big problems
with roachprod:
 - the out-of-memory behavior is very bad: instead of the process
   being killed, the system enters a thrashing mode where everything
   in the VM slows to a crawl (to the point where just sshing in can
   take minutes).
 - when the cockroach process exits, the exit code is not recorded
   anywhere, making it impossible in some cases to figure out why it
   stopped. In my particular case, we were exiting with exit code 8
   (which is `exit.TimeoutAfterFatalError()`) because writing to the
   logs was unacceptably slow.

This commit attempts to improve things on both these fronts. Instead
of running with `--background`, we use `systemd-run` to run cockroach
as a service unit. This has several advantages:
 - we have much better monitoring infrastructure via
   `systemctl status cockroach`
 - we can now run code after the exit, allowing us to record it in
   various logs.
 - we can set a strict cgroups memory limit (set to `95%`) so that the
   process gets oom-killed before the system starts to thrash.

As part of the commit, we also print out information about the status
of cockroach when logging in.

Fixes #64176.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64177)
<!-- Reviewable:end -->

RaduBerinde commented:
Hm, it looks like `tmux` is not available in the VMs in roachtest, even though it was there when I used roachprod create.

RaduBerinde commented:
I think a better way to do this would be to use `systemd-run` to run as a service, which would allow for better monitoring (including exit code). We can also apply a memory limit so that the process gets killed reliably when OOMing instead of the entire VM thrashing. I will play around with that.

tbg commented:
>  I will play around with that.

Cool, really appreciate it. I agree that it makes sense to productionize the ad-hoc starting/stopping we have right now. Let me know how I can help.

RaduBerinde commented:
Updated - we now run with `systemd-run` and apply a strict 95% memory limit.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> neat

> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-:-MZGoOJg-jymKkIF1YX9:bb76exl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, @irfansharif, and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
+cat > ${HOME}/run-cockroach.sh <<'EOF' && \
+sudo systemd-run --unit cockroach --uid $uid --gid $gid -p MemoryMax=%[9]s bash ${HOME}/run-cockroach.sh
+#!/bin/bash
+ulimit -c unlimited
```
no `set -euo pipefail`?

tbg commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
+# Output a message visible when checking the status (via sudo systemctl status).
+echo "Starting cockroach (output redirected to %[1]s)"
+export ROACHPROD=%[3]d%[4]s;
+set -x
```
do you want `-e` and `+e` here? `x` prints the commands the shell executes but that doesn't seem what you'd want here.

tbg commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
+code=$?
+set +x
+echo "cockroach exited with exit code $code" >> %[1]s/roachprod.log
+echo "cockroach exited with exit code $code" >> %[1]s/cockroach.stderr.log
```
consider throwing a $(date) in these so that we can figure out the exact time it stopped

tbg commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
+uid=$(id -u)
+gid=$(id -g)
+cat > ${HOME}/run-cockroach.sh <<'EOF' && \
+sudo systemd-run --unit cockroach --uid $uid --gid $gid -p MemoryMax=%[9]s bash ${HOME}/run-cockroach.sh
```
Some more useful flags that we may want to add
-r, --remain-after-exit
--working-directory=

I also wonder, don't we need some probe to determine when to return? `--background` returns when the ports are bound (so that you don't get spurious connection refused when connecting right after `c.Start()`). What prevents that flake here?

tbg replied:
```
--user

    Talk to the service manager of the calling user, rather than the service manager of the system.
```

Maybe this avoids the need for going through `root`?

RaduBerinde replied:
I tried that, the memory limit doesn't work with --user.

RaduBerinde replied:
Good question about the probe.. Are you sure it's not the same currently? AFAICT with `--background` we fork at the very beginning: https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/start.go#L283

tbg commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -396,19 +396,86 @@ func (h *crdbInstallHelper) generateStartCmd(
-	// NB: this is awkward as when the process fails, the test runner will show an
-	// unhelpful empty error (since everything has been redirected away). This is
-	// unfortunately equally awkward to address.
+	if h.c.IsLocal() {
```
We shouldn't let these diverge as much. Of course we can't ask for systemd to be installed locally, but can we share the `run-cockroach.sh` script and make the only difference that in `local` we invoke it with a `--background` arg directly?

cucaroach reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, and @RaduBerinde)

___
*[pkg/cmd/roachprod/install/cockroach.go, line 427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZISjjkAgbVKGLmlUCa:-MZISjjkAgbVKGLmlUCb:b9g72cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L427)):*
> ```Go
> 	// strict memory limit and provides tools to monitor the process.
> 	cmd := fmt.Sprintf(`
> sudo systemctl reset-failed cockroach
> ```

Is sudo guaranteed to work in all envs roachprod supports, is this a new requirement?
___
*[pkg/cmd/roachprod/install/cockroach.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZITYYU3zNI6tqiKGGV:-MZITYYU3zNI6tqiKGGW:b-raww0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L454)):*
> ```Go
> gid=$(id -g)
> cat > ${HOME}/run-cockroach.sh <<'EOF' && \
> sudo systemd-run --unit cockroach --uid $uid --gid $gid -p MemoryMax=%[9]s bash ${HOME}/run-cockroach.sh
> ```

I think we'll probably want the ability to play with MemoryMax % and also specify MemoryHigh which appears to be a "soft" limit (https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#MemoryHigh=bytes)


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-:-MZJ3eAlDZEk28WO7zDW:bww03zd)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @irfansharif, @RaduBerinde, and @tbg)

___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1Rxj3CsdDXlClsAP:-MZJ1Rxk97YKab4YLcjd:b-pyk1fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L159)):*
> ```Go
> }
> 
> // Stop TODO(peter): document
> ```

would you mind taking the opportunity and document the `wait`?
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1jgVES4bTI760fSc:-MZJ1jgVES4bTI760fSd:b5w6p3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L176)):*
> ```Go
> 			var resetServiceCmd string
> 			if !c.IsLocal() {
> 				// Since we're waiting (the default), clean up the service.
> ```

consider saying a few more words about what this `reset-failed` does.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ2Vnx3UEiCXna1ieC:-MZJ2Vnx3UEiCXna1ieD:b35tfup) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L451)):*
> ```Go
> fi
> 
> uid=$(id -u)
> ```

please write something about how `--uid` and `--gid` relate to `--user`, and how we can't use `--user`.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ2yyP5Y6wNFYMKzcx:-MZJ2yyP5Y6wNFYMKzcy:b-esx0by) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L461)):*
> ```Go
> echo ">>> roachprod start: $starttime" >> %[1]s/roachprod.log
> ps axeww -o pid -o command >> %[1]s/roachprod.log
> [ -x /usr/bin/lslocks ] && /usr/bin/lslocks >> %[1]s/roachprod.log
> ```

If you know why this is here, consider putting a comment.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTRs! I'll work more on it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-:-MZJS3ql8Le8JYRniAQQ:b-izb0d)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, @irfansharif, @RaduBerinde, and @tbg)

___
*[pkg/cmd/roachprod/install/cockroach.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-399:-MZJRa2j-_qy0r6728fS:b-1u8042) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L399)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We shouldn't let these diverge as much. Of course we can't ask for systemd to be installed locally, but can we share the `run-cockroach.sh` script and make the only difference that in `local` we invoke it with a `--background` arg directly?
</blockquote></details>

I don't want to create a script on your machine.. But I can extract the common stuff in the code here
___
*[pkg/cmd/roachprod/install/cockroach.go, line 427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZISjjkAgbVKGLmlUCa:-MZJS8mh1pidyQQb8sJl:b-5qt5kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L427)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Is sudo guaranteed to work in all envs roachprod supports, is this a new requirement?
</blockquote></details>

It's a new requirement, but it should be available in all VMs that we set up. We won't use it for "local" clusters.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZITYYU3zNI6tqiKGGV:-MZJSJcf3UlFAM6lNP0o:bwtms5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L454)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

I think we'll probably want the ability to play with MemoryMax % and also specify MemoryHigh which appears to be a "soft" limit (https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#MemoryHigh=bytes)
</blockquote></details>

We can play with that and add a config if we conclude it's useful. I am hesitant to use the soft limit (which slows down the program). It would be very hard to debug that slowdown, and it could be especially problematic with benchmarks.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ2yyP5Y6wNFYMKzcx:-MZJSgdg-SkhOOM1M3sv:bfb4k5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L461)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If you know why this is here, consider putting a comment.
</blockquote></details>

No idea, does anyone else?
___
*[pkg/cmd/roachprod/install/cockroach.go, line 468 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-468:-MZJRsW3D7cKAfywlkYO:b-5azkkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L468)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

do you want `-e` and `+e` here? `x` prints the commands the shell executes but that doesn't seem what you'd want here.
</blockquote></details>

We definitely don't want `-e`, we want to be able to run the code after it when it fails. I wanted to show the command so you can see it if you do `systemctl status cockroach`.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, @irfansharif, @RaduBerinde, and @tbg)

___
*[pkg/cmd/roachprod/install/cockroach.go, line 427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZISjjkAgbVKGLmlUCa:-MZJZpt_0V6SMIiS8dEa:btw8jio) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L427)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's a new requirement, but it should be available in all VMs that we set up. We won't use it for "local" clusters.
</blockquote></details>

We use `sudo` in `SyncedCluster.Wipe` and some other `SyncedCluster` operations.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/cmd/roachprod/install/cockroach.go, line 460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZKNFeq3uJi7WTnVi39:-MZKNFeq3uJi7WTnVi3A:bvlio89) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L460)):*
> ```Go
> starttime=$(date)
> echo ">>> roachprod start: $starttime" >> %[1]s/roachprod.log
> ps axeww -o pid -o command >> %[1]s/roachprod.log
> ```

Do we need these `ps axeww`s in the roachprod log? They are very unwieldy


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @andreimatei, @cucaroach, @irfansharif, and @tbg)

___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1Rxj3CsdDXlClsAP:-MZLEksI5ClcqQ6e_t2o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L159)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

would you mind taking the opportunity and document the `wait`?
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1jgVES4bTI760fSc:-MZLElHl0No4rXvPjPOA:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L176)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider saying a few more words about what this `reset-failed` does.
</blockquote></details>

Removed.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ2Vnx3UEiCXna1ieC:-MZLEnrKD710J0GpWrEH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L451)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please write something about how `--uid` and `--gid` relate to `--user`, and how we can't use `--user`.
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-456:-MZLEpASAs027s0ZlI4d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L456)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

no `set -euo pipefail`?
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 475 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-475:-MZLEu9E4ADcqPwMiTjp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L475)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

consider throwing a $(date) in these so that we can figure out the exact time it stopped
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @andreimatei, @cucaroach, @irfansharif, and @RaduBerinde)

___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1Rxj3CsdDXlClsAP:-MZMLr0qDbLfikPAmNnp:b-byc351) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done.
</blockquote></details>

But this takes the signal and wait, so there's no default here. Did you intend to put this comment elsewhere?
___
*[pkg/cmd/roachprod/install/cockroach.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-399:-MZMMAZz04aEhlFRGJ6J:b-wppe67) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L399)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't want to create a script on your machine.. But I can extract the common stuff in the code here
</blockquote></details>

Why not? I appreciate that you've introduced reuse here but... this stuff is just so hard to maintain or even read. I thought we would have a single script that simply knows whether it's invoked in local or remote mode. I have this PR checked out locally and it's so hard. I think we lost GOTRACEBACK=crash for the non-local case? We need to clean this up avoid the divergence between the paths or it'll be so hard to maintain and get right.

I spent some time on this - pushed a commit - pretty happy with how it turned out. We are of course pushing the boundaries here of what is reasonable but this is a lot less terrible than what we had before this PR.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-454:-MZMN03Y8NAVV1w_fH8M:b-b69qoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L454)):*
`--background` uses `sdnotify.Exec`:

https://github.com/cockroachdb/cockroach/blob/8d242779e742654034f11356f813cd19de857db7/pkg/cli/start_unix.go#L87

In other words, the parent waits until the child calls `sdnotify.Ready`, which it only does *after* binding the ports:

https://github.com/cockroachdb/cockroach/blob/6b1b69a4282ec1af0f1b94d61ba260073f3a00f1/pkg/cli/start.go#L422-L425

But I think this is... then already set up for success? We *are* running it under systemd now. The only question is, does systemd know that cockroach will signal systemd? It would have to decide to wait for the signal if things were to work out. I think that means using a service type of "Notify"?

https://askubuntu.com/questions/1120023/how-to-use-systemd-notify

> If you set up a service using Type=notify, systemd will automatically set up a communication socket back to systemd and will export its path to the service under $NOTIFY_SOCKET.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZITYYU3zNI6tqiKGGV:-MZMP3z5A9aMdCbL6ovO:bf6qozi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L454)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can play with that and add a config if we conclude it's useful. I am hesitant to use the soft limit (which slows down the program). It would be very hard to debug that slowdown, and it could be especially problematic with benchmarks.
</blockquote></details>

It's easy enough to play with them on a branch, I don't think we need to add knobs at this point.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZKNFeq3uJi7WTnVi39:-MZMPY2KAdCkGGgXmcRy:b6su03j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L460)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we need these `ps axeww`s in the roachprod log? They are very unwieldy
</blockquote></details>

No, they similarly come from past debugging. Let's remove.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ2yyP5Y6wNFYMKzcx:-MZMPN3HEZ1gO0sDrTf-:b-z34dkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L461)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

No idea, does anyone else?
</blockquote></details>

We sometimes run into issues in which `cockroach start` fails because .... well, we don't know, but one suspicion was that we were holding on to some RocksDB lock. This stuff hasn't been useful, let's remove. (Done, in my commit)
___
*[pkg/cmd/roachprod/install/cockroach.go, line 475 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-475:-MZMNVHi7-caXvoHsyn8:bb4pidv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L475)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done.
</blockquote></details>

That'll be in UTC, right?
Yep, just checked:

ubuntu@tobias-1619593141-08-n5cpu16-0001:~$ date
Wed Apr 28 07:29:57 UTC 2021


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/cmd/roachprod/install/cockroach.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-399:-MZNvAXH0d4jtKoWnBXk:b-rez0m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L399)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why not? I appreciate that you've introduced reuse here but... this stuff is just so hard to maintain or even read. I thought we would have a single script that simply knows whether it's invoked in local or remote mode. I have this PR checked out locally and it's so hard. I think we lost GOTRACEBACK=crash for the non-local case? We need to clean this up avoid the divergence between the paths or it'll be so hard to maintain and get right.

I spent some time on this - pushed a commit - pretty happy with how it turned out. We are of course pushing the boundaries here of what is reasonable but this is a lot less terrible than what we had before this PR.
</blockquote></details>

Very nice!!


<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -377,6 +378,83 @@ func (h *crdbInstallHelper) startNode(
+    background="--background"
+  fi
+  CODE=0
+  set +e
```
I don't think this is needed (because of the ||)

RaduBerinde reviewed (commented):
@tbg this looks good to me. If this looks good to you, I will squash and merge.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-:-MZPJViO1BrStuzQ_W2A:b-1yeund)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @andreimatei, @cucaroach, @irfansharif, @RaduBerinde, and @tbg)

___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZJ1Rxj3CsdDXlClsAP:-MZPJnKICUlq2uKHXaUM:b9dfuu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cluster_synced.go#L159)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

But this takes the signal and wait, so there's no default here. Did you intend to put this comment elsewhere?
</blockquote></details>

Rephrased.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r1-454:-MZPNgBq1iCv6_KzKDCI:br1c5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6297b2f457e6a30e487f2293ffec046f3ba193a/pkg/cmd/roachprod/install/cockroach.go#L454)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`--background` uses `sdnotify.Exec`:

https://github.com/cockroachdb/cockroach/blob/8d242779e742654034f11356f813cd19de857db7/pkg/cli/start_unix.go#L87

In other words, the parent waits until the child calls `sdnotify.Ready`, which it only does *after* binding the ports:

https://github.com/cockroachdb/cockroach/blob/6b1b69a4282ec1af0f1b94d61ba260073f3a00f1/pkg/cli/start.go#L422-L425

But I think this is... then already set up for success? We *are* running it under systemd now. The only question is, does systemd know that cockroach will signal systemd? It would have to decide to wait for the signal if things were to work out. I think that means using a service type of "Notify"?

https://askubuntu.com/questions/1120023/how-to-use-systemd-notify

> If you set up a service using Type=notify, systemd will automatically set up a communication socket back to systemd and will export its path to the service under $NOTIFY_SOCKET.
</blockquote></details>

Very cool! Yes, it works. I had to add `NotifyAccess=all` because the cockroach is a different PID than the one started by systemd (bash). I verified that it waits by adding a time.Sleep before calling sdnotify in the cockroach binary. In the meantime `systemctl status cockroach` shows it like this, very useful: `Active: activating (start) since Wed 2021-04-28 21:30:40 UTC; 8s ago`.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 391 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZPFnVQ31Iyd1AKogk9:-MZPFnVQ31Iyd1AKogkA:bogqxm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aa2c4c376221beb5b6b88330f3c4ca41c036be/pkg/cmd/roachprod/install/cockroach.go#L391)):*
> ```Go
> # \EOF ensures that no parameter expansion occurs in the heredoc. See:
> # https://unix.stackexchange.com/questions/376103/here-document-without-interpreting-escape-sequences-but-with-interpolation
> cat > "${helper}" << \EOF && chmod +x "${helper}" && "${helper}" "${verb}"
> ```

We are using `'EOF'` below which disables everything, made it the same.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 407 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-MZ3evlkvSe_iG8p2WEs-r4-407:-MZPJOnHB1av9RxA1Llg:b-en5eo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/11aa2c4c376221beb5b6b88330f3c4ca41c036be/pkg/cmd/roachprod/install/cockroach.go#L407)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think this is needed (because of the ||)
</blockquote></details>

Removed and it still works (verified that the error code makes it to roachprod.log)


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachprod: improve the way cockroach is run

While running some stress tests with TPCH, I observed two big problems
with roachprod:
 - the out-of-memory behavior is very bad: instead of the process
   being killed, the system enters a thrashing mode where everything
   in the VM slows to a crawl (to the point where just sshing in can
   take minutes).
 - when the cockroach process exits, the exit code is not recorded
   anywhere, making it impossible in some cases to figure out why it
   stopped. In my particular case, we were exiting with exit code 8
   (which is `exit.TimeoutAfterFatalError()`) because writing to the
   logs was unacceptably slow.

This commit attempts to improve things on both these fronts. Instead
of running with `--background`, we use `systemd-run` to run cockroach
as a service unit. This has several advantages:
 - we have much better monitoring infrastructure via
   `systemctl status cockroach`
 - we can now run code after the exit, allowing us to record it in
   various logs.
 - we can set a strict cgroups memory limit (set to `95%`) so that the
   process gets oom-killed before the system starts to thrash.

As part of the commit, we also print out information about the status
of cockroach when logging in.

Fixes #64176.

Release note: None

tbg reviewed (approved):
bors r+

:crossed_fingers: for the nightly run, I expect to see a few more OOMs on clusters that were previously just scraping by.

> Reviewed 1 of 3 files at r1, 2 of 3 files at r5, 1 of 1 files at r6.
> Dismissed @cucaroach, and @cucaroach from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64177#-:-MZROlxk2Mfasccrt6xY:b-ee34xm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei, @irfansharif, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
> for the nightly run, I expect to see a few more OOMs on clusters that were previously just scraping by.

Same here. I almost want to start a nightly run with a lower MemoryMax and adjust all the tests so they use less than 80-85%. If we let them scrape by, any small change can drive them over the limit.
